database.cc

```diff
@@ -2150,6 +2150,70 @@ static sstring get_snapshot_table_dir_prefix(const sstring& table_name) {
     return table_name + "-";
 }
 
+static sstring extract_cf_name(const sstring& directory_name) {
+    // cf directory is of the form: 'cf_name-timestamp'
+    return directory_name.substr(0, directory_name.find("-"));
+}
+
+future<std::vector<database::snapshot_details_result>>  database::get_snapshot_details() {
+    std::vector<sstring> data_dirs = _cfg.data_file_directories();
+    auto dirs_only_entries = lister::dir_entry_types{directory_entry_type::directory};
+    std::vector<database::snapshot_details_result> details;
+
+    for (auto& datadir : data_dirs) {
+        co_await lister::scan_dir(datadir, dirs_only_entries, [this, &dirs_only_entries, &details] (fs::path parent_dir, directory_entry de) -> future<> {
+            // KS directory
+            sstring ks_name = de.name;
+
+            co_return co_await lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &dirs_only_entries, &details, ks_name = std::move(ks_name)] (fs::path parent_dir, directory_entry de) -> future<> {
+                // CF directory
```

Comment:
It's orthogonal to the sstables version.
There shouldn't be any dependency even on scylla version at least form version 3.0 IIRC.

Reply:
I seem to have a vague recollection that the directory structure **was** different in old sstable versions - subdirectories were introduced for individual tables where previously all the tables were in the same per-keyspace directory. I'm not sure I'm remembering this right...

Reply:
Hmm, this is probably from the `ka` times apparently,
since they hold the keyspace and column_family name in the sstable name so they can all be placed in a shared data directory.

See https://github.com/scylladb/scylla/blob/master/docs/design-notes/sstables-directory-structure.md
```
    <keyspace>-<column_family>-ka-<generation>-<component>
```

Reply:
Right. So my question is - what does this patch do in this case? Does it at least do something reasonable (even if not perfect)?

Reply:
> Right (so I mixed up the details in my statement above). So my question is - what does this patch do in this case? Does it at least do something reasonable (even if not perfect)?

This patch or the present state before it AFAICT don't handle the shared-data-directory-with-no-pre-table-sub-directories case as they expect the snapshots subdirectory to live under the per-table directory.
Frankly, I don't know if we even supported snapshots in the shared data directory times...